File: be/src/main/java/sg/flow/controllers/TransferController.java
package sg.flow.controllers;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import lombok.RequiredArgsConstructor;
import sg.flow.entities.Bank;
import sg.flow.models.transfer.TransferRecepient;
import sg.flow.models.transfer.TransferRequestBody;
import sg.flow.models.transfer.TransferResult;
import sg.flow.services.TransferServices.TransferService;

import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestParam;

@RestController
@RequiredArgsConstructor
@RequestMapping("transaction/send")
public class TransferController {

    private final String SEND = "/send";

    @Autowired
    private final TransferService transferService;

    @GetMapping(SEND + "/getRelevantRecepient/accountNumber")
    public ResponseEntity<List<Bank>> getBanksByAccountNumber(
            @AuthenticationPrincipal(expression = "userId") Integer userId,
            @RequestParam(name = "keyword") String keyword) {
        return ResponseEntity.ok(transferService.getRelevantRecepientByAccountNumber(keyword));
    }

    @GetMapping(SEND + "/getRelevantRecepient/contact")
    public ResponseEntity<TransferRecepient> getRelevantRecepientByContact(
            @AuthenticationPrincipal(expression = "userId") Integer userId,
            @RequestParam(name = "keyword") String keyword) {
        return ResponseEntity.ok(transferService.getRelevantRecepientByContact(keyword));
    }

    @GetMapping(SEND + "/getRelevantRecepient")
    public ResponseEntity<List<TransferRecepient>> getRelevantRecepient(
            @AuthenticationPrincipal(expression = "userId") Integer userId) {
        return ResponseEntity.ok(transferService.getRelevantRecepient());
    }

    @PostMapping(SEND + "/sendTransaction")
    public ResponseEntity<TransferResult> sendTransaction(
            @AuthenticationPrincipal(expression = "userId") Integer userId, @RequestBody String jsonBody) {
        TransferRequestBody sendRequestBody = new TransferRequestBody(jsonBody);
        return ResponseEntity.ok(transferService.sendTransaction(sendRequestBody));
    }
}

File: be/src/main/kotlin/sg/flow/controllers/TransferController.kt
package sg.flow.controllers

import org.springframework.http.ResponseEntity
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.web.bind.annotation.*
import sg.flow.entities.Bank
import sg.flow.models.transfer.TransferRecepient
import sg.flow.models.transfer.TransferRequestBody
import sg.flow.models.transfer.TransferResult
import sg.flow.services.TransferServices.TransferService

@RestController
@RequestMapping("/transaction/send")
class TransferController(private val transferService: TransferService) {

    @GetMapping("/send/getRelevantRecepient/accountNumber")
    suspend fun getBanksByAccountNumber(
            @AuthenticationPrincipal(expression = "userId") userId: Int,
            @RequestParam(name = "keyword") keyword: String
    ): ResponseEntity<List<Bank>> =
            ResponseEntity.ok(transferService.getRelevantRecepientByAccountNumber(keyword))

    @GetMapping("/send/getRelevantRecepient/contact")
    suspend fun getRelevantRecepientByContact(
            @AuthenticationPrincipal(expression = "userId") userId: Int,
            @RequestParam(name = "keyword") keyword: String
    ): ResponseEntity<TransferRecepient> =
            ResponseEntity.ok(transferService.getRelevantRecepientByContact(keyword))

    @GetMapping("/send/getRelevantRecepient")
    suspend fun getRelevantRecepient(
            @AuthenticationPrincipal(expression = "userId") userId: Int
    ): ResponseEntity<List<TransferRecepient>> =
            ResponseEntity.ok(transferService.getRelevantRecepient())

    @PostMapping("/send/sendTransaction")
    suspend fun sendTransaction(
            @AuthenticationPrincipal(expression = "userId") userId: Int,
            @RequestBody sendRequestBody: TransferRequestBody
    ): ResponseEntity<TransferResult> =
            ResponseEntity.ok(transferService.sendTransaction(sendRequestBody))
}
